### PR TITLE
improve ZAP CRM exporting

### LINKED ISSUE #43: archive recoded ZAP CRM data before subsetting
* MapZAP uses the recoded ZAP CRM data archived by this repo
* that source data is currently limited to the subset of records and fields we publish to Open Data
* Planning Support would like to use other fields
* we likely want to improve the archiving process to recode and archive before subsetting

goals
- don't alter final open data output

approach
- build test process (on PRs: run open data export post-download, compare output to chosen expected output)
- start making changes

### PR DESCRIPTION
**Pull request by damonmcc**
resolves #43 

## changes
- reordering of ZAP CRM export logic to:
  - better separate logic of `runner.Runner` steps and helper functions
  - perform recoding of CRM data before subsetting to records marked for public visibility
  - export three versions of the data: `crm`, `internal`, `visible`
- expand sql engine class `pg.PG` to:
  - isolate all transactions to a specific schema in `edm-data.edm-zap` postgres database
  - more easily use parameters in sql queries
- add python integration tests to ensure main `runner.Runner` methods are working
- use a distinct postgres schema for all local and CI runs of tests and exports, e.g.:
  - `test_data_actual_damon`
  - `test_data_actual_pr_45`
  - `export_workflow_dispatch_improve_crm_export`
  - `export_schedule_main`
- use new exports as source data in MapZAP to include all records and more columns for CEQR Type II analysis
- cleanup/improve MapZAP dbt models

## notes
- there is now a sequence of tables built during CAP CRM export: `crm` -> 'staging' -> `recoded` -> `visible`
- depending on the dataset (aka whether any recoding is done or not), the export named `internal` may be a copy of either `crm` or `recoded`
- the new integration tests are primarily a first pass to ensure the function run without error. they do not yet compare outputs to expected records, but an approach to doing that has been started here and flagged with `TODO` codetags

## tests
- PR tests passing (unit and integration)
- [gha run](https://github.com/NYCPlanning/db-zap-opendata/actions/runs/5213279593) of export of all open datasets and a single non-open dataset (without `recode_id()` for `dcp_projects`)
- [gha run](https://github.com/NYCPlanning/db-zap-opendata/actions/runs/5215880304) of export of all datasets (in progress ...)

### CODE REVIEW

**Comment by damonmcc** [comment]
marking ready for review here. a full run of the export GHA [failed](https://github.com/NYCPlanning/db-zap-opendata/actions/runs/5214323395/jobs/9410450705#step:5:77) due to the dropping of the column `crm_project_id` during recode_id()

the commit https://github.com/NYCPlanning/db-zap-opendata/pull/45/commits/31ec7831f2f20df15de6416aacb9f89888812857 should ensure that column still doesn't appear in the open data output but is dropped later via the open data select statement

File: tests/test_runner.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+@pytest.mark.integration()
+@pytest.mark.parametrize("test_dataset", test_datasets)
+def test_runner_download(test_dataset):
+    runner = Runner(
+        name=test_dataset.table_name,
+        schema=test_schema_actual,
+    )
+    runner.download()
+    # TODO assert things
```
Comment (fvankrieken) [inline tests/test_runner.py:71]: Still have a todo here, not sure if you want in scope of this PR?

Comment (fvankrieken) [reply tests/test_runner.py:71]: Multiple below as well

**Comment by fvankrieken** [comment]
lgtm - nothing glaring. This is quite a codebase to come into, so props on figuring things out. Wish I could provide a little more comments on the meat of your changes but still working on my understanding of the codebase.

**Review by fvankrieken — APPROVED** [review]

**Comment by fvankrieken** [comment]
I actually like some of the OO stuff - for our python heavy workloads I honestly do like it for things like db drivers. But this repo leans way into it, and could use more comments or more verbose names, I had a hard time figuring out some of the things with the "recoding" for example. Outside of the scope of this PR though!

**Comment by damonmcc** [comment]
no worries it's definitely not easy to quickly understand how it all works (especially the recoding). and the diffs in this one are hard to follow too when it's a mix of big chunks of logic moving and changes within them

hope the PR description was helpful!

File: tests/test_runner.py (hunk 1 above)
Comment (damonmcc) [reply tests/test_runner.py:71]: yea those are our of scope for this one. I had started to do a little bit of it but decided to focus on "does it run"

maybe there's a better [codetag](https://peps.python.org/pep-0350/#mnemonics) than `TODO` for something like this? maybe `FIXME`?

Comment (fvankrieken) [reply tests/test_runner.py:71]: Nah TODO is to the point

Comment (fvankrieken) [reply tests/test_runner.py:71]: ... why do they call them mnemonics?

Comment (damonmcc) [reply tests/test_runner.py:71]: lol great question
